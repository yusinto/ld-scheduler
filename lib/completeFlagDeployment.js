'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _constants = require('./constants');

var _getRequestHeaders = require('./getRequestHeaders');

var _getRequestHeaders2 = _interopRequireDefault(_getRequestHeaders);

var _without = require('lodash/without');

var _without2 = _interopRequireDefault(_without);

var _log = require('./log');

var _log2 = _interopRequireDefault(_log);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

var log = new _log2.default('completeFlagDeployment');

exports.default = function () {
  var _ref = _asyncToGenerator(regeneratorRuntime.mark(function _callee(_ref2, environment, apiKey) {
    var key = _ref2.key,
        tags = _ref2.tags,
        description = _ref2.description;
    var updatedTags, operations, remainingScheduledTags, body, url, response;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            updatedTags = (0, _without2.default)(tags, environment + '-scheduled');
            operations = [{
              op: 'replace',
              path: '/tags',
              value: updatedTags
            }];
            remainingScheduledTags = updatedTags.filter(function (tag) {
              return tag.endsWith('-scheduled');
            });


            if (remainingScheduledTags.length === 0) {
              operations.push({
                op: 'replace',
                path: '/description',
                value: description
              });
            }

            body = JSON.stringify(operations);
            url = _constants.launchDarklyFlagsEndpoint + '/' + key;
            _context.prev = 6;
            _context.next = 9;
            return fetch(url, {
              method: 'PATCH',
              headers: (0, _getRequestHeaders2.default)(apiKey),
              body: body
            });

          case 9:
            response = _context.sent;


            log.info('LaunchDarkly api response: ' + response.status + ' ' + response.statusText + ' from: ' + response.url);

            if (response.status === 200) {
              log.info('SUCCESS LD api! Updated ' + key + ' to ' + body + '.');
            } else {
              log.error('LaunchDarkly threw an error. Did not update ' + key + '. Will retry again later.');
            }
            _context.next = 17;
            break;

          case 14:
            _context.prev = 14;
            _context.t0 = _context['catch'](6);

            log.error('Network error. Could not reach LaunchDarkly. Did not update ' + key + '. Will retry again later. ' + _context.t0);

          case 17:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, undefined, [[6, 14]]);
  }));

  return function (_x, _x2, _x3) {
    return _ref.apply(this, arguments);
  };
}();