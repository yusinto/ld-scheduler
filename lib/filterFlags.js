'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = filterFlags;

var _isValidDateAfter = require('./isValidDateAfter');

var _isValidDateAfter2 = _interopRequireDefault(_isValidDateAfter);

var _sendSlackMessgae = require('./sendSlackMessgae');

var _sendSlackMessgae2 = _interopRequireDefault(_sendSlackMessgae);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function filterFlags(scheduledFlags, slack) {
  var description = '';
  // get only flags that can be deployed
  return scheduledFlags.reduce(function (accumulator, f) {
    try {
      description = JSON.parse(f.description);
    } catch (e) {
      var message = f.key + ' is scheduled, but its description field is not a valid json object: ' + f.description;
      return (0, _sendSlackMessgae2.default)(message, slack);
    }

    return Array.isArray(description) ? [].concat(_toConsumableArray(accumulator), _toConsumableArray(description.map(function (value) {
      if (!(0, _isValidDateAfter2.default)(value)) return null;

      log.info('Found scheduled flag ' + f.key);
      return _extends({
        key: f.key,
        tags: f.tags
      }, value, {
        originalDescription: description
      });
    }))) : [].concat(_toConsumableArray(accumulator), [_extends({
      key: f.key,
      tags: f.tags
    }, description)]);
  }, []).filter(Boolean);
}