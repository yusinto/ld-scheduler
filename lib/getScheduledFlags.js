'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _config = require('config');

var _config2 = _interopRequireDefault(_config);

var _getRequestHeaders = require('./getRequestHeaders');

var _getRequestHeaders2 = _interopRequireDefault(_getRequestHeaders);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

// list all scheduled flags in the default project
exports.default = function () {
  var _ref = _asyncToGenerator(regeneratorRuntime.mark(function _callee(ldEnvironment, apiKey) {
    var url, response, data;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            url = '' + _config2.default.launchDarkly.rest.baseUrl + _config2.default.launchDarkly.rest.flags + '?env=' + ldEnvironment + '&tag=scheduled';
            _context.prev = 1;
            _context.next = 4;
            return fetch(url, {
              method: 'GET',
              headers: (0, _getRequestHeaders2.default)(apiKey)
            });

          case 4:
            response = _context.sent;

            if (!(response.status === 200)) {
              _context.next = 10;
              break;
            }

            _context.next = 8;
            return response.json();

          case 8:
            data = _context.sent;
            return _context.abrupt('return', data.items);

          case 10:

            console.log('getScheduledFlag ERROR: api response: ' + response.status + ' ' + response.statusText + ' from: ' + response.url);
            return _context.abrupt('return', []);

          case 14:
            _context.prev = 14;
            _context.t0 = _context['catch'](1);

            console.log('getScheduledFlags EXCEPTION: ' + _context.t0 + '. Will retry again later.');
            return _context.abrupt('return', []);

          case 18:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, undefined, [[1, 14]]);
  }));

  return function (_x, _x2) {
    return _ref.apply(this, arguments);
  };
}();