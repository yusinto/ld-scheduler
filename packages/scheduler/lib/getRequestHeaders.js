'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

exports.default = function (apiKey) {
  return {
    Accept: '*/*',
    'Content-Type': 'application/json',
    Authorization: apiKey,
    'accept-encoding': 'gzip, deflate'
  };
};