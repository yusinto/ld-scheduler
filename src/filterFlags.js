import isValidDateAfter from './isValidDateAfter';
import sendSlackMessage from './sendSlackMessgae';

export default function filterFlags(scheduledFlags, slack) {
  let description = '';
  // get only flags that can be deployed
  return scheduledFlags
    .reduce((accumulator, f) => {
      try {
        description = JSON.parse(f.description);
      } catch (e) {
        const message = `${f.key} is scheduled, but its description field is not a valid json object: ${f.description}`;
        return sendSlackMessage(message, slack);
      }

      return Array.isArray(description) ? [
        ...accumulator,
        ...description.map(value => {
          if (!isValidDateAfter(value)) return null;

          log.info(`Found scheduled flag ${f.key}`);
          return {
            key: f.key,
            tags: f.tags,
            ...value,
            originalDescription: description,
          };
        }),
      ] : [
        ...accumulator,
        {
          key: f.key,
          tags: f.tags,
          ...description,
        },
      ];
    }, []).filter(Boolean);
}
