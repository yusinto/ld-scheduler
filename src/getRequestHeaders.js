export default (apiKey) => ({
  Accept: '*/*',
  'Content-Type': 'application/json',
  Authorization: apiKey,
  'accept-encoding': 'gzip, deflate'
});
