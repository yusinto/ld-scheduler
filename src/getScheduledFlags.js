import config from 'config';
import {requestHeaders} from 'constants';

// list all scheduled flags in the default project
export default getScheduledFlags = async() => {
  const url = `${config.launchDarkly.rest.baseUrl}${config.launchDarkly.rest.flags}?env=${config.launchDarkly.environment}&tag=scheduled`;
  try {
    const response = await fetch(url, {
      method: 'GET',
      headers: requestHeaders,
    });

    const data = await response.json();
    return data.items;
  } catch (e) {
    console.log(`getScheduledFlags: ERROR: ${e}. Will retry again later.`);
    return [];
  }
};