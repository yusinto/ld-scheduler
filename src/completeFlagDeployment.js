import {launchDarklyFlagsEndpoint} from './constants';
import getRequestHeaders from './getRequestHeaders';
import without from 'lodash/without';
import Logger from './log';

const log = new Logger('completeFlagDeployment');

export default async({key, tags, description, targetDeploymentDateTime, originalDescription}, environment, apiKey) => {
  const updatedTags = without(tags, `${environment}-scheduled`);
  const operations = [
    {
      op: 'replace',
      path: '/tags',
      value: updatedTags,
    },
  ];

  const commonAttributes = {
    op: 'replace',
    path: '/description',
  };

  Array.isArray(originalDescription) ?
    operations.push({
      ...commonAttributes,
      value: JSON.stringify([
        ...originalDescription.map((d) => {
          console.log(d);
          return {
            ...d,
            ...d.targetDeploymentDateTime === targetDeploymentDateTime ? { __isDeployed: true } : null,
          };
        }),
      ]),
    }) :
    operations.push({
      ...commonAttributes,
      value: JSON.stringify({
        ...description,
        __isDeployed: true,
      }),
    });

  const body = JSON.stringify(operations);
  const url = `${launchDarklyFlagsEndpoint}/${key}`;

  try {
    const response = await fetch(url, {
      method: 'PATCH',
      headers: getRequestHeaders(apiKey),
      body,
    });

    log.info(`LaunchDarkly api response: ${response.status} ${response.statusText} from: ${response.url}`);

    if (response.status === 200) {
      log.info(`SUCCESS LD api! Updated ${key} to ${body}.`);
    } else {
      log.error(`LaunchDarkly threw an error. Did not update ${key}. Will retry again later.`);
    }
  } catch (e) {
    log.error(`Network error. Could not reach LaunchDarkly. Did not update ${key}. Will retry again later. ${e}`);
  }
};
