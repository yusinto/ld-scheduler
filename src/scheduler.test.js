jest.mock('moment', () => global.td.function('moment'));
jest.mock('./constants', () => ({
  taskTypes: {
    killSwitch: 'killSwitch',
    fallThroughRollout: 'fallThroughRollout'
  },
  launchDarklyFlagsEndpoint: 'mockBaseUrl/flags'
}));
jest.mock('./getRequestHeaders', () => global.td.function('getRequestHeaders'));
jest.mock('./getScheduledFlags', () => global.td.function('getScheduledFlags'));
jest.mock('./completeFlagDeployment', () => global.td.function('completeFlagDeployment'));
jest.mock('./slack', () => global.td.function('messageSlack'));

import moment from 'moment';
import getRequestHeaders from './getRequestHeaders';
import getScheduledFlags from './getScheduledFlags';
import completeFlagDeployment from './completeFlagDeployment';
import messageSlack from './slack';
import filterFlags from './filterFlags';


const mockMomentFormat = td.function('mockMomentFormat');
const mockMomentIsAfter = td.function('mockMomentIsAfter');

describe('Scheduler', () => {
  const scheduler = require('./scheduler').default;

  beforeEach(async() => {
    const momentObject = {
      format: mockMomentFormat,
      isAfter: mockMomentIsAfter
    };

    td.when(moment()).thenReturn(momentObject);
    td.when(moment(td.matchers.anything(), 'YYYY-MM-DD HH:mm Z')).thenReturn(momentObject);
    td.when(mockMomentFormat(td.matchers.anything())).thenReturn('TestDate');
    td.when(getRequestHeaders(td.matchers.anything())).thenReturn('headers');
  });

  afterEach(() => {
    jest.resetAllMocks();
    td.reset();
  });

  it('does not do anything if there is no outstanding task', async() => {
    fetch.mockSuccess();
    td.when(mockMomentIsAfter(td.matchers.anything())).thenReturn(false);
    td.when(getScheduledFlags(td.matchers.anything(), td.matchers.anything())).thenReturn([]);

    await scheduler({environment: 'test', apiKey: 'someKey', slack: '/url/to/slack/webhook'});

    expect(fetch).not.toHaveBeenCalledWith('/url/to/slack/webhook', {
      method: 'POST',
      body: JSON.stringify({text: '[Test] Successfully switched on test-flag.'})
    });
  });

  it('patches when there is a scheduled killSwitch flag', async() => {
    // arrange
    fetch.mockSuccess();
    const scheduledFlag = {
      key: 'flag1',
      tags: ['scheduled', 'some other tag'],
      description: `{
          "taskType": "killSwitch",
          "value": true,
          "targetDeploymentDateTime": "2017-02-27 22:00", 
          "description": "Flag1 description"
        }`
    };
    const outstandingTask = {
      key: scheduledFlag.key,
      tags: scheduledFlag.tags,
      ...JSON.parse(scheduledFlag.description)
    };
    td.when(getScheduledFlags(td.matchers.anything(), td.matchers.anything())).thenReturn([scheduledFlag]);
    td.when(mockMomentIsAfter(td.matchers.anything())).thenReturn(true);

    // act
    await scheduler({environment: 'test', apiKey: 'someKey', slack: '/url/to/slack/webhook'});

    // assert
    expect(fetch).toHaveBeenCalledWith('mockBaseUrl/flags/flag1', {
      method: 'PATCH',
      headers: 'headers',
      body: JSON.stringify([{
        op: 'replace',
        path: '/environments/test/on',
        value: true,
      }])
    });
    td.verify(completeFlagDeployment(outstandingTask, 'test', 'someKey'));
    td.verify(messageSlack({isUpdateSuccessful: true, task: outstandingTask}, 'test', '/url/to/slack/webhook'));
  });

  it('patches when there is a scheduled fallThroughRollout flag', async() => {
    // arrange
    fetch.mockSuccess();
    const scheduledFlag = {
      key: 'flag1',
      tags: ['scheduled', 'some other tag'],
      description: `{
        "taskType": "fallThroughRollout",
        "targetDeploymentDateTime": "2017-03-30 02:33", 
        "description": "Flag1 description",
        "value": [
           {
             "variation": 0,
             "weight": 100000
           },
           {
             "variation": 1,
             "weight": 0
           }
         ]
      }`
    };
    const outstandingTask = {
      key: scheduledFlag.key,
      tags: scheduledFlag.tags,
      ...JSON.parse(scheduledFlag.description)
    };
    td.when(getScheduledFlags(td.matchers.anything(), td.matchers.anything())).thenReturn([scheduledFlag]);
    td.when(mockMomentIsAfter(td.matchers.anything())).thenReturn(true);

    // act
    await scheduler({environment: 'test', apiKey: 'someKey', slack: '/url/to/slack/webhook'});

    // assert
    expect(fetch).toHaveBeenCalledWith('mockBaseUrl/flags/flag1', {
      method: 'PATCH',
      headers: 'headers',
      body: JSON.stringify([{
        op: 'replace',
        path: '/environments/test/fallthrough/rollout/variations',
        value: [
          {
            variation: 0,
            weight: 100000
          },
          {
            variation: 1,
            weight: 0
          }
        ],
      }])
    });
    td.verify(completeFlagDeployment(outstandingTask, 'test', 'someKey'));
    td.verify(messageSlack({isUpdateSuccessful: true, task: outstandingTask}, 'test', '/url/to/slack/webhook'));
  });

  it('updates slack with correct message if path fails', async() => {
    // arrange
    fetch.mockSuccess({}, {status: 401, statusText: 'Unauthorized'});
    const scheduledFlag = {
      key: 'flag1',
      tags: ['scheduled', 'some other tag'],
      description: `{
        "taskType": "fallThroughRollout",
        "targetDeploymentDateTime": "2017-03-30 02:33", 
        "description": "Flag1 description",
        "value": [
           {
             "variation": 0,
             "weight": 100000
           },
           {
             "variation": 1,
             "weight": 0
           }
         ]
      }`
    };
    const outstandingTask = {
      key: scheduledFlag.key,
      tags: scheduledFlag.tags,
      ...JSON.parse(scheduledFlag.description)
    };
    td.when(getScheduledFlags(td.matchers.anything(), td.matchers.anything())).thenReturn([scheduledFlag]);
    td.when(mockMomentIsAfter(td.matchers.anything())).thenReturn(true);

    // act
    await scheduler({environment: 'test', apiKey: 'someKey', slack: '/url/to/slack/webhook'});

    // assert
    expect(fetch).toHaveBeenCalledWith('mockBaseUrl/flags/flag1', {
      method: 'PATCH',
      headers: 'headers',
      body: JSON.stringify([{
        op: 'replace',
        path: '/environments/test/fallthrough/rollout/variations',
        value: [
          {
            variation: 0,
            weight: 100000
          },
          {
            variation: 1,
            weight: 0
          }
        ],
      }])
    });
    td.verify(completeFlagDeployment(outstandingTask, 'someKey'), {times: 0});
    td.verify(messageSlack({isUpdateSuccessful: false, task: outstandingTask}, 'test', '/url/to/slack/webhook'));
  });

  describe('filterFlags', () => {
    it('should render correct for json object as string', () => {
      const scheduledFlag = [{
        key: 'flag1',
        tags: ['scheduled', 'some other tag'],
        description: `{
          "taskType": "fallThroughRollout",
          "targetDeploymentDateTime": "2017-03-30 02:33", 
          "description": "Flag1 description",
          "value": "test"
        }`
      }];
      td.when(mockMomentIsAfter(td.matchers.anything())).thenReturn(true);
      console.log(filterFlags(scheduledFlag));
      expect(filterFlags(scheduledFlag)).toMatchSnapshot();
    });

    it('should render correct for array as string', () => {
      const scheduledFlag = [{
        key: 'flag1',
        tags: ['scheduled', 'some other tag'],
        description: `[
          {
            "taskType": "fallThroughRollout",
            "targetDeploymentDateTime": "2017-03-30 02:33", 
            "description": "Flag1 description",
            "value": "test1"
          },
          {
            "taskType": "fallThroughRollout1",
            "targetDeploymentDateTime": "2017-03-31 02:33", 
            "description": "Flag1 description",
            "value": "test2"
          }
        ]`
      }];
      td.when(mockMomentIsAfter(td.matchers.anything())).thenReturn(true);
      expect(filterFlags(scheduledFlag)).toMatchSnapshot();
    });
  });
});
