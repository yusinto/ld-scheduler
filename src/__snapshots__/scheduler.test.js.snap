// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scheduler filterRequiredFilters should render correct for array as string 1`] = `
Array [
  Object {
    "description": "Flag1 description",
    "key": "flag1",
    "originalDescription": Array [
      Object {
        "description": "Flag1 description",
        "targetDeploymentDateTime": "2017-03-30 02:33",
        "taskType": "fallThroughRollout",
        "value": "test1",
      },
      Object {
        "description": "Flag1 description",
        "targetDeploymentDateTime": "2017-03-31 02:33",
        "taskType": "fallThroughRollout1",
        "value": "test2",
      },
    ],
    "tags": Array [
      "scheduled",
      "some other tag",
    ],
    "targetDeploymentDateTime": "2017-03-30 02:33",
    "taskType": "fallThroughRollout",
    "value": "test1",
  },
  Object {
    "description": "Flag1 description",
    "key": "flag1",
    "originalDescription": Array [
      Object {
        "description": "Flag1 description",
        "targetDeploymentDateTime": "2017-03-30 02:33",
        "taskType": "fallThroughRollout",
        "value": "test1",
      },
      Object {
        "description": "Flag1 description",
        "targetDeploymentDateTime": "2017-03-31 02:33",
        "taskType": "fallThroughRollout1",
        "value": "test2",
      },
    ],
    "tags": Array [
      "scheduled",
      "some other tag",
    ],
    "targetDeploymentDateTime": "2017-03-31 02:33",
    "taskType": "fallThroughRollout1",
    "value": "test2",
  },
]
`;

exports[`Scheduler filterRequiredFilters should render correct for json object as string 1`] = `
Array [
  Object {
    "description": "Flag1 description",
    "key": "flag1",
    "tags": Array [
      "scheduled",
      "some other tag",
    ],
    "targetDeploymentDateTime": "2017-03-30 02:33",
    "taskType": "fallThroughRollout",
    "value": "test",
  },
]
`;
